### PR TITLE
Refactor path storage

### LINKED ISSUE #4071: Refactor handling of entry point in goto-symex
entry point and resume need to be handled uniformly

and clean up get_goto_functions: https://github.com/diffblue/cbmc/pull/3969#discussion_r253271211

### PR DESCRIPTION
**Pull request by peterschrammel**
Removes 70 lines and a lot of complexity from the single-path symex checker.

Fixes #4071

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 792d854c2eecf79655d33ea934e6f4280d2499ba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100012351

File: src/goto-symex/path_storage.cpp [hunk 1]
```diff
@@ -29,12 +29,9 @@ path_storaget::patht &path_lifot::private_peek()
   return paths.back();
 }
 
-void path_lifot::push(
-  const path_storaget::patht &next_instruction,
-  const path_storaget::patht &jump_target)
+void path_lifot::push(const path_storaget::patht &path)
```
Comment (tautschnig) [inline src/goto-symex/path_storage.cpp:32]: I don't think we actually need these overrides anymore.

File: src/goto-symex/path_storage.h [hunk 2]
```diff
@@ -68,16 +68,8 @@ class path_storaget
...
-  /// Symbolic execution is suspended when we reach a conditional goto
-  /// instruction with two successors. Thus, we always add a pair of paths to
-  /// the path storage.
-  ///
-  /// \param next_instruction: the instruction following the goto instruction
-  /// \param jump_target: the target instruction of the goto instruction
-  virtual void
-  push(const patht &next_instruction, const patht &jump_target) = 0;
+  /// \brief Add a path to resume to the storage
+  virtual void push(const patht &) = 0;
```
Comment (tautschnig) [inline src/goto-symex/path_storage.h:72]: I'd suggest to not make this one abstract - all current implementations are the same, so we might as well provide a default implementation. Future deriving classes are still free to override.

File: src/goto-checker/single_path_symex_only_checker.cpp [hunk 3]
```diff
@@ -37,29 +37,24 @@ operator()(propertiest &properties)
 {
   resultt result(resultt::progresst::DONE);
 
-  // First run goto-symex from entry point to initialize worklist
+  auto get_goto_function =
+    [this](const irep_idt &id) -> const goto_functionst::goto_functiont & {
+    return goto_model.get_goto_function(id);
+  };
```
Comment (tautschnig) [inline src/goto-checker/single_path_symex_only_checker.cpp]: Could this also be refactored? Code copies are undesirable...

File: src/goto-checker/multi_path_symex_checker.cpp [hunk 4]
```diff
@@ -41,14 +41,13 @@ operator()(propertiest &properties)
...
-      symex,
-      symex_symbol_table,
-      equation,
-      options,
-      ns,
-      ui_message_handler);
+    auto get_goto_function =
+      [this](const irep_idt &id) -> const goto_functionst::goto_functiont & {
+      return goto_model.get_goto_function(id);
+    };
```
Comment (tautschnig) [inline src/goto-checker/multi_path_symex_checker.cpp]: The third copy?

File: src/goto-checker/multi_path_symex_only_checker.cpp [hunk 5]
```diff
@@ -39,14 +39,13 @@ multi_path_symex_only_checkert::multi_path_symex_only_checkert(
...
-    symex,
-    symex_symbol_table,
-    equation,
-    options,
-    ns,
-    ui_message_handler);
+  auto get_goto_function =
+    [this](const irep_idt &id) -> const goto_functionst::goto_functiont & {
+    return goto_model.get_goto_function(id);
+  };
```
Comment (tautschnig) [inline src/goto-checker/multi_path_symex_only_checker.cpp]: Copy nr#4

**Review by tautschnig — APPROVED** [review]

File: src/goto-symex/path_storage.cpp (hunk 1 above)
Comment (peterschrammel) [reply src/goto-symex/path_storage.cpp:32]: The interface doesn't define the storage used.

File: src/goto-symex/path_storage.h (hunk 2 above)
Comment (peterschrammel) [reply src/goto-symex/path_storage.h:72]: They would also have to inherit unused internal storage.

File: src/goto-checker/single_path_symex_only_checker.cpp (hunk 3 above)
Comment (peterschrammel) [reply src/goto-checker/single_path_symex_only_checker.cpp]: Yes... unfortunately, C++ lambdas are so clumsy that a one-line lambda feels like code duplication...

File: src/goto-symex/path_storage.h (hunk 2 above)
Comment (tautschnig) [reply src/goto-symex/path_storage.h:72]: Ah, sorry for missing that.

File: src/goto-checker/single_path_symex_only_checker.cpp (hunk 3 above)
Comment (tautschnig) [reply src/goto-checker/single_path_symex_only_checker.cpp]: Shouldn't we really have an interface that doesn't require passing this parameter? I believe that in fact we already do, in `goto_symext::symex_with_state`, which has a variant with and without that function. So maybe we could also do that for those other functions?

Comment (peterschrammel) [reply src/goto-checker/single_path_symex_only_checker.cpp]: > variant without that function

That one should actually be removed.

Comment (tautschnig) [reply src/goto-checker/single_path_symex_only_checker.cpp]: Ok, then maybe that could happen here as well? I dislike code copies, but I even more dislike almost-code-copies which we currently have with `get_function_from_goto_functions`.

Comment (peterschrammel) [reply src/goto-checker/single_path_symex_only_checker.cpp]: Check the changes, please.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5281fa654b9674517fa6b9aef67c4060de6a77d8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100093034

**Review by chrisr-diffblue — APPROVED** [review]
Looks good to me.

**Pull request merged**